File: packages/python/catalogmx/catalogs/sat/carta_porte/aeropuertos.py
"""Catálogo c_CodigoTransporteAereo - Aeropuertos"""

import json
from pathlib import Path

from catalogmx.utils.text import normalize_text


class AeropuertosCatalog:
    _data: list[dict] | None = None
    _by_code: dict[str, dict] | None = None
    _by_iata: dict[str, dict] | None = None
    _by_icao: dict[str, dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        if cls._data is None:
            path = (
                Path(__file__).parent.parent.parent.parent.parent.parent
                / "shared-data"
                / "sat"
                / "carta_porte_3"
                / "aeropuertos.json"
            )
            with open(path, encoding="utf-8") as f:
                cls._data = json.load(f)
            cls._by_code = {item["code"]: item for item in cls._data}
            cls._by_iata = {item["iata"]: item for item in cls._data}
            cls._by_icao = {item["icao"]: item for item in cls._data}

    @classmethod
    def get_aeropuerto(cls, code: str) -> dict | None:
        """Obtiene aeropuerto por código SAT"""
        cls._load_data()
        return cls._by_code.get(code)

    @classmethod
    def get_by_iata(cls, iata: str) -> dict | None:
        """Obtiene aeropuerto por código IATA"""
        cls._load_data()
        return cls._by_iata.get(iata)

    @classmethod
    def get_by_icao(cls, icao: str) -> dict | None:
        """Obtiene aeropuerto por código ICAO"""
        cls._load_data()
        return cls._by_icao.get(icao)

    @classmethod
    def is_valid(cls, code: str) -> bool:
        """Verifica si un código de aeropuerto es válido"""
        return cls.get_aeropuerto(code) is not None

    @classmethod
    def get_all(cls) -> list[dict]:
        """Obtiene todos los aeropuertos"""
        cls._load_data()
        return cls._data.copy()

    @classmethod
    def get_by_state(cls, state: str) -> list[dict]:
        """Obtiene aeropuertos por estado (insensible a acentos)"""
        cls._load_data()
        state_normalized = normalize_text(state)
        return [
            a
            for a in cls._data
            if normalize_text(a.get("estado", a.get("state", ""))) == state_normalized
        ]

    @classmethod
    def search_by_name(cls, name: str) -> list[dict]:
        """Busca aeropuertos por nombre (insensible a acentos)"""
        cls._load_data()
        name_normalized = normalize_text(name)
        return [a for a in cls._data if name_normalized in normalize_text(a["name"])]

File: packages/typescript/src/catalogs/sat/carta_porte/aeropuertos.ts
/**
 * SAT Carta Porte 3.0 - Aeropuertos
 * Mexican airports with IATA and ICAO codes
 */

import { loadCatalogObject } from '../../../utils/catalog-loader';
import type { Aeropuerto } from '../../../types';

export class AeropuertosCatalog {
  private static _data: Aeropuerto[] | null = null;

  private static getData(): Aeropuerto[] {
    if (!this._data) {
      this._data = loadCatalogObject<Aeropuerto>('sat/carta_porte_3/aeropuertos.json');
    }
    return this._data;
  }

  static getAll(): Aeropuerto[] {
    return this.getData();
  }

  static getByCode(code: string): Aeropuerto | undefined {
    return this.getData().find((a) => a.code === code.toUpperCase());
  }

  static getByIATA(iata: string): Aeropuerto | undefined {
    return this.getData().find((a) => a.iata === iata.toUpperCase());
  }

  static getByICAO(icao: string): Aeropuerto | undefined {
    return this.getData().find((a) => a.icao === icao.toUpperCase());
  }

  static isValid(code: string): boolean {
    return this.getData().some((a) => a.code === code.toUpperCase());
  }

  static isValidIATA(iata: string): boolean {
    return this.getData().some((a) => a.iata === iata.toUpperCase());
  }

  static isValidICAO(icao: string): boolean {
    return this.getData().some((a) => a.icao === icao.toUpperCase());
  }

  /**
   * Get airports by state
   */
  static getByEstado(estado: string): Aeropuerto[] {
    const search = estado.toUpperCase();
    return this.getData().filter((a) => a.estado.toUpperCase().includes(search));
  }

  /**
   * Get airports by city
   */
  static getByCiudad(ciudad: string): Aeropuerto[] {
    const search = ciudad.toUpperCase();
    return this.getData().filter((a) => a.ciudad.toUpperCase().includes(search));
  }

  /**
   * Search airports by name
   */
  static searchByName(keyword: string): Aeropuerto[] {
    const search = keyword.toUpperCase();
    return this.getData().filter(
      (a) => a.name.toUpperCase().includes(search) || a.ciudad.toUpperCase().includes(search)
    );
  }
}
